logger/proto.go

```diff
@@ -0,0 +1,135 @@
+	m protoreflect.Map
+}
+
+func (p protoMapMarshaller) MarshalLogObject(e zapcore.ObjectEncoder) error {
+	p.m.Range(func(ki protoreflect.MapKey, vi protoreflect.Value) bool {
+		var k string
+		switch p.f.MapKey().Kind() {
+		case protoreflect.BoolKind:
+			k = strconv.FormatBool(ki.Bool())
+		case protoreflect.Int32Kind, protoreflect.Int64Kind, protoreflect.Sint32Kind, protoreflect.Sint64Kind, protoreflect.Sfixed32Kind, protoreflect.Sfixed64Kind:
+			k = strconv.FormatInt(ki.Int(), 10)
+		case protoreflect.Uint32Kind, protoreflect.Uint64Kind, protoreflect.Fixed32Kind, protoreflect.Fixed64Kind:
+			k = strconv.FormatUint(ki.Uint(), 10)
+		case protoreflect.StringKind:
+			k = ki.String()
+		}
+		marshalProtoField(k, p.f.MapValue(), vi, e)
+		return true
+	})
+	return nil
+}
+
+type protoListMarshaller struct {
+	f protoreflect.FieldDescriptor
+	m protoreflect.List
+}
+
+func (p protoListMarshaller) MarshalLogArray(e zapcore.ArrayEncoder) error {
+	for i := 0; i < p.m.Len(); i++ {
+		v := p.m.Get(i)
+		switch p.f.Kind() {
+		case protoreflect.BoolKind:
+			e.AppendBool(v.Bool())
+		case protoreflect.EnumKind:
+			e.AppendInt32(int32(v.Enum()))
+		case protoreflect.Int32Kind, protoreflect.Int64Kind, protoreflect.Sint32Kind, protoreflect.Sint64Kind, protoreflect.Sfixed32Kind, protoreflect.Sfixed64Kind:
+			e.AppendInt64(v.Int())
+		case protoreflect.Uint32Kind, protoreflect.Uint64Kind, protoreflect.Fixed32Kind, protoreflect.Fixed64Kind:
+			e.AppendUint64(v.Uint())
+		case protoreflect.FloatKind, protoreflect.DoubleKind:
+			e.AppendFloat64(v.Float())
+		case protoreflect.StringKind:
+			e.AppendString(v.String())
+		case protoreflect.BytesKind:
+			e.AppendString(base64.RawStdEncoding.EncodeToString(v.Bytes()))
+		case protoreflect.MessageKind:
+			e.AppendObject(protoMarshaller{v.Message()})
+		}
+	}
+	return nil
+}
+
+func marshalProtoField(k string, f protoreflect.FieldDescriptor, v protoreflect.Value, e zapcore.ObjectEncoder) {
+	switch f.Kind() {
+	case protoreflect.BoolKind:
+		if b := v.Bool(); b {
+			e.AddBool(k, b)
+		}
+	case protoreflect.EnumKind:
+		if n := v.Enum(); n != 0 {
+			e.AddInt32(k, int32(n))
+		}
+	case protoreflect.Int32Kind, protoreflect.Int64Kind, protoreflect.Sint32Kind, protoreflect.Sint64Kind, protoreflect.Sfixed32Kind, protoreflect.Sfixed64Kind:
+		if n := v.Int(); n != 0 {
+			e.AddInt64(k, n)
+		}
+	case protoreflect.Uint32Kind, protoreflect.Uint64Kind, protoreflect.Fixed32Kind, protoreflect.Fixed64Kind:
+		if n := v.Uint(); n != 0 {
+			e.AddUint64(k, n)
+		}
+	case protoreflect.FloatKind, protoreflect.DoubleKind:
+		if n := v.Float(); n != 0 {
+			e.AddFloat64(k, n)
+		}
+	case protoreflect.StringKind:
+		if s := v.String(); s != "" {
+			e.AddString(k, s)
+		}
+	case protoreflect.BytesKind:
+		if b := v.Bytes(); len(b) != 0 {
+			e.AddString(k, base64.RawStdEncoding.EncodeToString(b))
+		}
+	case protoreflect.MessageKind:
+		if m := v.Message(); m.IsValid() {
+			e.AddObject(k, protoMarshaller{m})
+		}
```

Comment:
Will these kinds change (more like augmented) with protobuf version? Is there a need to marshal a generic placeholder "unknown kind"?

Reply:
i don't think these change in major pb versions